### PR TITLE
[FIX] microsoft_calendar: add condition on check

### PR DESCRIPTION
**Pull request by ryce-odoo**
Added condition on check that ensured the organizer was an attendee on the event on create. This was added in a change that allowed the organizer to be changed on the Odoo side. However, this caused issues with the appointments app when creating an appointment that only used resources. The "organizer" who in this case is the creator of the appointment type would not be an attendee and therefore would cause an error and make the appointment type unbookable.

Adding this condition allows the appointment to bypass the check if the appointment type uses resources instead of users.

opw-3841495

### CODE REVIEW

**Review by geflx — COMMENTED** [review]
Hello! :wave: Thanks for your work :smile: 
I left some comments and feedback on the PR.
Also, does this problem starts from `17.4` or does it happen in any branch before as well?

**Comment by ryce-odoo** [comment]
Hey @geflx 
This is actually just a forward port in advance for another PR that you actually helped me review before. Renaud just wanted me to make a PR for this version because it might affect a change that was made in 17.3 and beyond.
https://github.com/odoo/odoo/pull/162927

File: addons/microsoft_calendar/models/calendar.py [hunk 1]
```diff
@@ -73,6 +73,8 @@ def _check_microsoft_sync_status(self):
     def _skip_send_mail_status_update(self):
         """If microsoft calendar is not syncing, don't send a mail."""
         user_id = self._get_event_user_m()
+        if user_id.partner_id.id not in self.partner_ids.ids:
```
Comment (reth-odoo) [inline addons/microsoft_calendar/models/calendar.py:76]: Don't think we expect newids here so this should be fine (and faster)
```suggestion
        if user_id.partner_id not in self.partner_ids:
```

File: addons/calendar/models/calendar_event.py [hunk 2]
```diff
@@ -431,6 +431,12 @@ def _check_closing_date(self):
                     )
                 )
 
+    def _check_organizer_validation_conditions(self, vals_list):
+        """ Method for check in the microsoft_calendar module that needs to be
+            overridden in appointment.
+        """
```
[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/calendar/models/calendar_event.py:437]: See PR linked below, sadly there's no bridge for appointment and microsoft_calendar so we had to add the method in base calendar

As it's effectively only used in microsoft_calendar I think it may be worth mentionning

Also thanks for having a look!

File: addons/microsoft_calendar/models/calendar.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ def _skip_send_mail_status_update(self):
         """If microsoft calendar is not syncing, don't send a mail."""
         user_id = self._get_event_user_m()
         if self.with_user(user_id)._check_microsoft_sync_status() and user_id._get_microsoft_sync_status() == "sync_active":
-            return True
+            return not (self.user_id and self.user_id.partner_id not in self.partner_ids)
```
Comment (reth-odoo) [inline addons/microsoft_calendar/models/calendar.py:77]: Would prefer avoiding double negations as it's a bit confusing imo, isn't this condition sufficient?
"Skip emails if the organizer is an attendee"
`self.user_id.partner_id in self.partner_ids`

**Review by reth-odoo — COMMENTED** [review]
Ok, microsoft sync with appointment seems broken already i.e. not considering we will sync if the current user isn't the organizer, which is never the case for portal.

I'll make a separate fix for that but for now this fix seems fine and should prevent tracebacks. Just one comment

File: addons/microsoft_calendar/models/calendar.py (hunk 3 above)
Comment (ryce-odoo) [reply addons/microsoft_calendar/models/calendar.py]: That would make sense! Thanks

**Comment by ryce-odoo** [comment]
@reth-odoo Sounds good! I have made that change so this should be ready whenever.

**Comment by reth-odoo** [comment]
@ryce-odoo at least one relevant test is failing, and apparently needs a rebase

**Comment by ryce-odoo** [comment]
@reth-odoo I have rebased, but I wonder why that test is failing because it wasn't failing before I will look into it maybe the simplified logic doesn't work as we would expect.

**Comment by reth-odoo** [comment]
@ryce-odoo isn't it expected? We don't sync anymore for resources so we send an email? Maybe you can just update the test

Would be nice to close this one, seems like it's still broken in this version :/

**Comment by reth-odoo** [comment]
@ryce-odoo isn't the whole point of the validation that it cannot sync without the organizer being an attendee?

I just tried in 17.0 and it does not sync.

**Comment by ryce-odoo** [comment]
@reth-odoo That is what I thought too, but maybe something changed in this time? Because it is still syncing even when the organizer is not an attendee in this test or at least a mock_insert is called. I distinctly remember running these new test cases locally when I first made these changes and everything worked fine, but that was months ago, but when I pushed the changes here it didn't work anymore.

**Comment by reth-odoo** [comment]
@ryce-odoo ok then, I'll just pool it in task 4146850 if I notice it's a problem.

Thanks

**Comment by reth-odoo** [comment]
I think I misunderstood, so the fix is still needed

I'll mess around with the tests for a bit then for the "skip email" part

**Comment by reth-odoo** [comment]
@ryce-odoo ok yes indeed, the sync either works if the user even if the organizer isn't in the attendees, or `_check_microsoft_sync_status` will fail and the email will be sent.

I took the liberty to write a test to ensure the behavior in appointment tests, and to rebase.
Otherwise looks good to me then, indeed it could probably have gone on as a regular forward-port back then if I'd realized this :/

**Comment by reth-odoo** [comment]
@jeh-odoo if you have some time to check this one it's an old forward port of a ticket fix (https://github.com/odoo/odoo/pull/162927) that completely blocks booking resource appointments when the original creator of the appointment type is syncing their microsoft calendar

In the end not much changes from the original and it's been working fine in 17.0 all this time so should be a fairly safe merge

File: addons/microsoft_calendar/models/microsoft_sync.py [hunk 4]
```diff
@@ -87,15 +87,14 @@ def write(self, vals):
...
         records = super().create(vals_list)
 
-        if self.env.user._get_microsoft_sync_status() != "sync_paused":
-            for record in records:
-                if record.need_sync_m and record.active:
-                    record._microsoft_insert(record._microsoft_values(self._get_microsoft_synced_fields()), timeout=3)
+        for record in records:
+            user_id = record._get_event_user_m()
+            if user_id.microsoft_synchronization_stopped:
+                record.need_sync_m = False
```
[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py:95]: @geflx  hey, thanks for having a look

I can add a test specifically for this in community but to make sure I get this right, does the test in appointment on the related branch cover this?

I create the event manually and check that either insert is called once or it creates a mail (granted it's an appointment event but it shouldn't change much)

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py:95]: @geflx I meant more `test_sync_or_email_resource_appointment` actually :D

I'll partially move it in here already anyway we can see after if it's ok with you

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py:95]: @geflx Here you go, managed to have a fairly minimal test in the end. Just checks this case in an existing test case.

Let me know if there's some edge case (maybe past the original create?) you can think of but otherwise seems to work fine. At least functionally it works good when testing manually, but I didn't really check for staggered double syncing or anything like that.

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py:95]: The issue here is that in the very specific case where:
- You have paused your microsoft syncing
- The organizer is syncing

`_skip_send_mail_status_update` sees the microsoft sync user (the organizer) is syncing, so it doesn't send the mail
And `create` sees the current user is NOT syncing, so it doesn't sync even though it would have for a write call for example

I can fix it in `_skip_send_mail_status_update` instead if you think this is too touchy, I can understand not wanting to update this just for a weird edge case

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py]: I ended up fixing the `_skip_send_mail` method to just send a mail if the event was not marked to be synced nor was synced. That seems like the best option in the end. At worst if we miss something we'll send emails when it's not necessary but that's a lot better than not sending emails nor syncing...

If you can have a quick look-over to see if it holds water @geflx

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py]: @geflx :up: :arrow_up:

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/microsoft_calendar/models/microsoft_sync.py]: @geflx :cry:

**Review by geflx — APPROVED** [review]
Thank you for your patience, I agree with the changes.

**Review by reth-odoo — COMMENTED** [review]
No worries thanks for the review, just wanted sure this wasn't forgotten again :D

@jeh-odoo you can finally have a look at this one when you have time (also the sibling in enterprise appointment), as said above it entirely prevents booking in some rare cases so would be nice to close (workarounds are to not sync with outlook or have some other user to create the appointment types...)

(pushed to rebase)

**Review by jeh-odoo — APPROVED** [review]

**Comment by reth-odoo** [comment]
@arj-odoo / @geflx  Hi :wave: do you mind r+-ing this one, as it's technically a forward-port of this older one which you r+-ed (had it checked by jeh for appointment side) https://github.com/odoo/odoo/pull/162927

Cheers

**Comment by arj-odoo** [comment]
@robodoo r+ rebase-ff
